src/bigistore.c

```diff
@@ -978,9 +866,352 @@ bigistore_svals(PG_FUNCTION_ARGS)
+        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("array must not contain nulls")));
+
+    pairs   = FIRST_PAIR(is, BigIStorePair);
+    creator = palloc0(sizeof *creator);
+
+    bigistore_pairs_init(creator, MIN(is->len, alen));
+
+    if (alen > 1)
+        qsort((void *) ar, alen, sizeof(int32), is_int32_arr_comp);
+
+    while (index1 < is->len && index2 < alen)
+    {
+        if (pairs[index1].key < ar[index2])
+        {
+            ++index1;
+        }
+        else if (pairs[index1].key > ar[index2])
+        {
+            ++index2;
+        }
+        else
+        {
+            bigistore_pairs_insert(creator, pairs[index1].key, pairs[index1].val);
+            ++index1;
+            ++index2;
+            found = true;
+        }
+    }
+    if (found)
+    {
+        FINALIZE_BIGISTORE(result, creator);
+        PG_RETURN_POINTER(result);
+    }
+    else
+        PG_RETURN_NULL();
+}
+
+PG_FUNCTION_INFO_V1(bigistore_slice_to_array);
+Datum bigistore_slice_to_array(PG_FUNCTION_ARGS)
+{
+    BigIStorePair *pair;
+    BigIStore *    is   = PG_GETARG_BIGIS(0);
+    ArrayType *    a    = PG_GETARG_ARRAYTYPE_P(1);
+    int32 *        ar   = (int32 *) ARR_DATA_PTR(a);
+    int            alen = ArrayGetNItems(ARR_NDIM(a), ARR_DIMS(a));
+    int            lbs[1];
+    int            dims[1];
+    int            i;
+    Datum *        out_datums;
+    bool *         out_nulls;
+    ArrayType *    aout;
+
+    if (ARR_HASNULL(a) && array_contains_nulls(a))
+        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("array must not contain nulls")));
+
+    out_datums = palloc(sizeof(Datum) * alen);
+    out_nulls  = palloc(sizeof(bool) * alen);
+    dims[0]    = alen;
+    lbs[0]     = 1;
+
+    for (i = 0; i < alen; ++i)
+    {
+        if ((pair = bigistore_find(is, ar[i], NULL)) == NULL)
+        {
+            out_datums[i] = (Datum) 0;
+            out_nulls[i]  = true;
+        }
+        else
+        {
+            out_datums[i] = pair->val;
+            out_nulls[i]  = false;
+        }
+    }
+
+    aout = construct_md_array(out_datums, out_nulls, 1, dims, lbs, INT8OID, sizeof(int64), true, 'i');
+
+    PG_RETURN_POINTER(aout);
+}
+
+PG_FUNCTION_INFO_V1(bigistore_delete);
+Datum bigistore_delete(PG_FUNCTION_ARGS)
+{
+    BigIStorePair *pair;
+    BigIStore *    is  = PG_GETARG_BIGIS_COPY(0);
+    int32          key = PG_GETARG_INT32(1);
+    int            del_off;
+
+    if ((pair = bigistore_find(is, key, &del_off)))
+    {
+        --is->len;
+        is->buflen -= bigis_pair_buf_len(pair);
+        if (is->len > del_off)
+            memmove(pair, pair + 1, (is->len - del_off) * sizeof(BigIStorePair));
+    }
+    SET_VARSIZE(is, ISHDRSZ + (is->len * sizeof(BigIStorePair)));
+    PG_RETURN_POINTER(is);
+}
+
+PG_FUNCTION_INFO_V1(bigistore_delete_array);
+Datum bigistore_delete_array(PG_FUNCTION_ARGS)
+{
+    BigIStorePair *pairs;
+    BigIStore *    is     = PG_GETARG_BIGIS_COPY(0);
```

Comment:
did you validate the performance argument?
my experience is different

Reply:
Sure, naive pairwise copying would be slower than `memcpy` + `memove` even if they are performed multiple times. Esp. if there are just a few keys to delete. (I actually measured today and it actually is, not by a huge margin though). I want to sketch contiguous-block-wise one-pass copying, that should be faster.

Reply:
_I've initially posted this in the wrong thread._

So, something like a snippet below is slightly faster in my unscientific test and, IMO, is a bit more straightforward: copying "surviving" elements into a new place in blocks instead of copying everything and then iteratively collapsing it. Although, despite more straightforward logic, it might be even harder to read.

```C
PG_FUNCTION_INFO_V1(istore_delete_array);
Datum istore_delete_array(PG_FUNCTION_ARGS)
{
    IStorePair * pairs, * dst_pairs;
    IStore     * result;
    IStore     * is      = PG_GETARG_IS(0);
    ArrayType  * a       = PG_GETARG_ARRAYTYPE_P_COPY(1);
    int32      * ar      = (int32 *) ARR_DATA_PTR(a);
    int          alen    = ArrayGetNItems(ARR_NDIM(a), ARR_DIMS(a));
    int          is_kept_offset = 0, ar_index = 0, is_left = 0;

    if (ARR_HASNULL(a) && array_contains_nulls(a))
        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("array must not contain nulls")));

    if (alen > 1)
        qsort((void *) ar, alen, sizeof(int32), is_int32_arr_comp);

    is_left   = is->len;
    pairs     = FIRST_PAIR(is, IStorePair);
    result    = palloc0(ISTORE_SIZE(is, IStorePair));
    SET_VARSIZE(result, ISTORE_SIZE(is, IStorePair));
    dst_pairs = FIRST_PAIR(result, IStorePair);

    while ((is_left - is_kept_offset > 0) && ar_index < alen)
    {
        if (pairs[is_kept_offset].key < ar[ar_index])
        {
            ++is_kept_offset;
        }
        else if (pairs[is_kept_offset].key > ar[ar_index])
        {
            ++ar_index;
        }
        else
        {
            if (is_kept_offset > 0) 
            {
                memcpy(dst_pairs, pairs, is_kept_offset * sizeof(IStorePair));

                dst_pairs   += is_kept_offset;
                result->len += is_kept_offset;

                pairs       += is_kept_offset;
                is_left     -= is_kept_offset;

                is_kept_offset = 0;
            } 

            ++pairs;
            --is_left;
            ++ar_index;
        }
    }
    if (is_left > 0)
    {
        memcpy(dst_pairs, pairs, is_left * sizeof(IStorePair));
        result->len += is_left;
    }

    result = repalloc(result, ISTORE_SIZE(result, IStorePair));
    SET_VARSIZE(result, ISTORE_SIZE(result, IStorePair));

    PG_RETURN_POINTER(result);
}
```

And here is the SQL snippet I used for speed comparison:
```SQL
with
 keys as (
	select s1 as id, array_agg(s2) as ks 
	from generate_series(1, 1000) s1, generate_series(3, 1000, 6) s2
	group by s1
),
 istores as (
	select s1 as id, sum(istore(ARRAY[s1, s2])) as ist
	from generate_series(1, 1000) s1, generate_series(0, 2000, 3) s2
	group by s1
)
select max(istore_length(delete(ist, ks))) 
from istores join keys on istores.id = keys.id;
```

Reply:
One more version, less code than the other two, same performance.  Iw as expecting it to be faster in "small array, big istore" cases, but wasn't able to find any evidence of it. Probably the runtime in those cases is dominated by something else.

```C
PG_FUNCTION_INFO_V1(istore_delete_array);
Datum istore_delete_array(PG_FUNCTION_ARGS)
{
    IStorePair *pairs;
    IStore *    is   = PG_GETARG_IS_COPY(0);
    ArrayType * a    = PG_GETARG_ARRAYTYPE_P(1);
    int32 *     ar   = (int32 *) ARR_DATA_PTR(a);
    int         alen = ArrayGetNItems(ARR_NDIM(a), ARR_DIMS(a));
    int         i = 0, is_i = 0;

    if (ARR_HASNULL(a) && array_contains_nulls(a))
        ereport(ERROR, (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED), errmsg("array must not contain nulls")));

    pairs = FIRST_PAIR(is, IStorePair);

    for (i = 0; i < alen; i++)
    {
        if (istore_find(is, ar[i], &is_i) != NULL)
        {
            --is->len;
            is->buflen -= is_pair_buf_len(pairs + is_i);
            memmove(pairs + is_i, pairs + is_i + 1, (is->len - is_i) * sizeof(IStorePair));
        }
    }

    SET_VARSIZE(is, ISTORE_SIZE(is, IStorePair));
    PG_RETURN_POINTER(is);
}
```